ENH: stats: Add a function that computes the relative risk.

WarrenWeckesser opened the pull request:
The new function `scipy.stats.contingency.relative_risk` computes the relative
risk (also known as the risk ratio) for a 2x2 contingency table.
The object returned by the function has a method to compute the
confidence interval of the relative risk using a standard, widely
used method.

WarrenWeckesser commented:
Here's an example where we can use the function to reproduce the calculations shown at https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717_ComparingFrequencies/PH717_ComparingFrequencies7.html:

The contingency table is

           |  CVD | No CVD | Total
    -------|------|--------|------
    HTN    |  992 |   2260 |  3252
    No HTN |  165 |   1017 |  1182
    Total  | 1157 |   3277 |  4434

The "No HTN" row corresponds to the "control" group.  The "CVD" column holds the counts of the "cases".  So the data to pass to `relative_risk` is

```
In [10]: exposed_cases = 992 
    ...: exposed_total = 3252 
    ...: control_cases = 165  
    ...: control_total = 1182                                                                                   

In [11]: result = relative_risk(exposed_cases, exposed_total, control_cases, control_total)                     

In [12]: result.relative_risk                                                                                   
Out[12]: 2.1852174885385214

In [13]: result.confidence_interval()                                                                           
Out[13]: ConfidenceInterval(low=1.879441362833471, high=2.5407419282373804)
```
The web site reports a relative risk of 2.18, with 95% confidence interval (1.87, 2.53).  The slight discrepancy is the result of the digits of intermediate results being truncated during the calculation shown on that web site.

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+                 control_cases, control_total,
+                 relative_risk):
+        self.relative_risk = relative_risk
+        self.exposed_cases = exposed_cases
+        self.exposed_total = exposed_total
+        self.control_cases = control_cases
+        self.control_total = control_total
```
should these check `>0`? otherwise will be division by `0` in the `log(rr)` calculation, unless I missed something.

WarrenWeckesser replied:
Good point.  I'll add checks for 0 and handle it appropriately.

WarrenWeckesser replied:
Done.

WarrenWeckesser replied:
The latest commit is `f56eea7...`:
```
In [5]: relative_risk(1, 0, 1, 1)                                               
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-c4e8ffb0b537> in <module>
----> 1 relative_risk(1, 0, 1, 1)

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+f56eea7-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_relative_risk.py in relative_risk(exposed_cases, exposed_total, control_cases, control_total)
    234 
    235     exposed_cases = _validate_int(exposed_cases, 0, "exposed_cases")
--> 236     exposed_total = _validate_int(exposed_total, 1, "exposed_total")
    237     control_cases = _validate_int(control_cases, 0, "control_cases")
    238     control_total = _validate_int(control_total, 1, "control_total")

~/mc37allpip/lib/python3.7/site-packages/scipy-1.6.0.dev0+f56eea7-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_relative_risk.py in _validate_int(n, bound, name)
     13         raise TypeError(msg) from None
     14     if n < bound:
---> 15         raise ValueError(msg)
     16     return n
     17 

ValueError: exposed_total must be an integer not less than 1, but got 0
```

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+
+        alpha = 1 - confidence_level
+        z = ndtri(1 - alpha/2)
+        rr = self.relative_risk
```
this one should maybe also have a check on non-negativity (wherever other checks wind up)-seeing after looking at the calculation a second time-right?

WarrenWeckesser replied:
`self.relative_risk` is computed by the main function.  If I ensure that the inputs are nonnegative then `self.relative_risk` will not be negative.  Is that what you are concerned about?

It is possible for `self.relative_risk` to be 0, inf or nan, and the code that computes the confidence interval should handle these cases "nicely".  I'll figure out what "nicely" means as I update the code to be more careful about the edge cases. :)

WarrenWeckesser replied:
Done.  I used the same convention as used by `riskratio` in the R library `epitools`.

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+    """
+    # This is a trivial calculation.  The nontrivial part is in the
+    # `confidence_interval` method of the RelativeRiskResult class.
+    if ((exposed_cases == 0 and exposed_total == 0) or
```
Maybe check `nan` and `inf` and throw value error on those cases?

WarrenWeckesser replied:
The values are the elements of a contingency table, so we can require them to be nonnegative integers.  I'll add some code to validate the values.

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/stats/tests/test_relative_risk.py:
```diff
@@ -0,0 +1,46 @@
+    assert_allclose(result.relative_risk, expected_rr, rtol=1e-13)
+
+
+def test_relative_risk_confidence_interval():
```
because this is (in some sense) decoupled from the `relative_risk` we may want tests for `0` counts as well as `nan`/`inf`.

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/stats/tests/test_relative_risk.py:
```diff
@@ -0,0 +1,46 @@
+    result = relative_risk(exposed_cases, exposed_total,
+                           control_cases, control_total)
+    rr = result.relative_risk
+    ci = result.confidence_interval(confidence_level=0.95)
```
maybe add test for `confidence_level=1` edge case ?

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+    exposed_cases : int
+        The number of "cases" (i.e. occurrence of disease or other event
+        of interest) among the sample of "exposed" individuals.
+    exposed_total : int
+        The total number of "exposed" individuals in the sample.
```
not sure if this is the best spot in docstring but maybe mention the `>0` s?

WarrenWeckesser replied:
I updated the docstring to note the constraints on those inputs.

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+ConfidenceInterval = namedtuple('ConfidenceInterval', ['low', 'high'])
+
+
+class RelativeRiskResult:
```
does this need a docstring?

WarrenWeckesser replied:
Yup, done.

mdhaber replied:
Should it appear in the documentation? I can't find it now.

WarrenWeckesser replied:
At the bottom of the [main stats page](https://24801-1460385-gh.circle-artifacts.com/0/html-scipyorg/stats.html) is a link to [`scipy.stats.contingency.RelativeRiskResult`](https://24801-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.stats.contingency.RelativeRiskResult.html#scipy.stats.contingency.RelativeRiskResult)

WarrenWeckesser replied:
...but I pushed a change that move the location of the link up to just below the list of `contingency` functions.

rlucas7 reviewed (changes requested):
@WarrenWeckesser I left a few comments, mostly around edge cases, clarifying docstrings.

WarrenWeckesser commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,197 @@
+        """
+        if not 0 <= confidence_level <= 1:
+            raise ValueError('confidence_level must be in the interval '
+                             '[0, 1].')
```
I think it works as expected.

As the confidence level decreases, the confidence interval shrinks:
```
In [28]: result = relative_risk(10, 100, 15, 350)                                                           

In [29]: result.confidence_interval(0.95)                                                                   
Out[29]: ConfidenceInterval(low=1.0818021000414402, high=5.032754553014722)

In [30]: result.confidence_interval(0.9)                                                                    
Out[30]: ConfidenceInterval(low=1.2241050406176928, high=4.447693836549465)

In [31]: result.confidence_interval(0.8)                                                                    
Out[31]: ConfidenceInterval(low=1.4115548468524273, high=3.8570548332463352)
```
As the confidence level approaches 0, the bounds of the confidence interval converge to the estimator:
```
In [33]: result.confidence_interval(0.1)                                                                    
Out[33]: ConfidenceInterval(low=2.221128262304914, high=2.4512066848380134)

In [34]: result.confidence_interval(1e-3)                                                                   
Out[34]: ConfidenceInterval(low=2.3321867075033142, high=2.3344805229050074)

In [35]: result.confidence_interval(0)                                                                      
Out[35]: ConfidenceInterval(low=2.3333333333333335, high=2.3333333333333335)

In [36]: result.relative_risk                                                                               
Out[36]: 2.3333333333333335
```
As the confidence level increases, the confidence interval grows.
```
In [37]: result.confidence_interval(0.95)                                                                   
Out[37]: ConfidenceInterval(low=1.0818021000414402, high=5.032754553014722)

In [38]: result.confidence_interval(0.975)                                                                  
Out[38]: ConfidenceInterval(low=0.9687507219628377, high=5.620067496221489)

In [39]: result.confidence_interval(0.999)                                                                  
Out[39]: ConfidenceInterval(low=0.6419792787760221, high=8.480716783919654)

In [40]: result.confidence_interval(0.999999)                                                               
Out[40]: ConfidenceInterval(low=0.3426203901425983, high=15.890602547555538)

In [41]: result.confidence_interval(0.999999999)                                                            
Out[41]: ConfidenceInterval(low=0.21251883114006848, high=25.61864478191149)
```
In the limit of a confidence level of 1, the confidence interval covers the entire domain of the estimator:
```
In [42]: result.confidence_interval(1)                                                                      
Out[42]: ConfidenceInterval(low=0.0, high=inf)
```
That says "we are 100% confident that the relative risk is in the interval [0, inf]".  Not very useful, but correct.

WarrenWeckesser replied:
I added unit tests for these limits of the confidence level.

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,260 @@
+
+    To pass the same data to ``relative_risk``, use::
+
+        relative_risk(n11, n10 + n11, n01, n00 + n01)
```
just a comment: 
when you come back to this one:
https://github.com/scipy/scipy/pull/11352/files
it might make sense to add the `table` as a see also here, or maybe even a small tutorial page, it does seem like the two would play well together :)

rlucas7 reviewed (changes requested):
@WarrenWeckesser thanks for changes and added tests, looks pretty close. 

I left a couple questions, more so than changes, I could see them stay as is or change, all minor. 

Only question I still  had was about the testing for `>0` in the `__init__` of `RelativeRiskResult`, there might still be a check that got missed?

Other than getting these last changes for the `>0` it's good from my end.

WarrenWeckesser commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,260 @@
+from scipy.special import ndtri
+
+
+def _validate_int(n, bound, name):
```
For now, let's leave it here.  We can move it to `_lib` whenever it turns out to be useful to do so.

mdhaber replied:
I will probably move it over shortly.

WarrenWeckesser commented on scipy/stats/tests/test_relative_risk.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def test_relative_risk_bad_type():
+    with pytest.raises(TypeError, match="must be an integer"):
```
Agreed.  I'd rather err on the side of stronger typing, and relax it later if there is sufficient justification.

rlucas7 replied:
ok, marking resolved.

WarrenWeckesser commented on scipy/stats/tests/test_relative_risk.py:
```diff
@@ -0,0 +1,97 @@
+    exposed_cases = 16
+    exposed_total = 128
+    control_cases = 24
+    control_total = 256
```
Agreed.  I just pushed an update to fix this (`f779a4e...`).

rlucas7 replied:
mark resolved

rlucas7 reviewed (approved):
Thanks @WarrenWeckesser these changes look good to me and I now am seeing the `>0` checks-not sure why they didn't show up in my branch checkout from remote yesterday evening. 

I'll mark approve and leave up for a few days and merge on Sunday Nov 15th if no more comments.

rgommers reviewed (changes requested):
It looks to me like this should go into `stats.contingency`. Adding relatively obscure functions that only apply to 2x2 contingency tables doesn't seem quite right - we have a special namespace for it after all.

rgommers commented:
I think I also missed a mailing list message about this new feature, can you send one @WarrenWeckesser?

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,260 @@
+
+import operator
```
```suggestion
from operator import index
```
I think that's all that's being used here, right?

WarrenWeckesser replied:
Yes, but we almost always use the fully qualified name `operator.index` elsewhere.  I think that is preferable, because `index` sounds fairly generic, and almost looks like it could be a numpy function.

mdhaber replied:
I'd agree with that.

rlucas7 replied:
ok, marking resolved

rlucas7 commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,260 @@
+def _validate_int(n, bound, name):
+    msg = f'{name} must be an integer not less than {bound}, but got {n!r}'
+    try:
+        n = operator.index(n)
```
```suggestion
        n = index(n)
```
if you agree w/limiting the import I'm suggesting above.

rlucas7 reviewed (changes requested):
@WarrenWeckesser 

I left two (very minor) suggested 1-liner changes 

I also agree w/Ralf's request, shouldn't this be put into the `scipy/stats/contingency.py`
module rather than top level `scipy.stats`?

Both of those seem like small changes so I'll remove my approval and move back to request changes (sorry for stepping off but it does seem minor and the right place to put and a small change).

WarrenWeckesser commented:
I pushed an update:

* Move `relative_risk` to the `scipy.stats.contingency` namespace.
* Add a second reference in the `relative_risk` docstring.
* Make `ConfidenceInterval` a class instead of a namedtuple (cf. https://github.com/scipy/scipy/pull/12603)

WarrenWeckesser commented:
I pushed a change to fix the refguide check failure.

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+
+        # Handle edge cases where either exposed_cases or control_cases
+        # is zero.  We follow the convention of the R function riskratio
+        # from the epitools library.
```
These seem like sensible conventions.

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+        """
+        Compute the confidence interval for the relative risk.
+
+        The confidence interval is computed using the Katz method
```
There are three methods in Katz' paper. Maybe mention which this is? C, I think?

WarrenWeckesser replied:
Done.

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+        #                1/control_cases - 1/control_total
+        # and the standard error is the square root of that.
+        se = np.sqrt(1/self.exposed_cases - 1/self.exposed_total +
+                     1/self.control_cases - 1/self.control_total)
```
Is this how it is presented in the "Statistics in Epidemiology" reference, or have you manipulated it for accuracy/speed?

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+        delta = z*se
+        katz_lo = rr*np.exp(-delta)
+        katz_hi = rr*np.exp(delta)
+        return ConfidenceInterval(low=katz_lo, high=katz_hi)
```
Implementation reflects the math. I'd just suggest making it explicit that this is Method C of the Katz paper.

WarrenWeckesser replied:
Done.

mdhaber replied:
Where is this done? I don't see it.

WarrenWeckesser replied:
It is in the docstring of the `confidence_interval` method of the `RelativeRiskResult` class.

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+            return ConfidenceInterval(low=np.nan, high=np.inf)
+
+        alpha = 1 - confidence_level
+        z = ndtri(1 - alpha/2)
```
I assume that the convention Katz follows for defining confidence limits is different, as naiive implementation of Katz would be `(1 - alpha)`. Might be worth noting. I assume this is because Katz seems to think of the upper and lower confident limits separately, each with confidence (1-alpha), whereas we think of a confidence interval with confidence (1-alpha)?

WarrenWeckesser replied:
This gives the interval around the mean of a normal distribution, i.e. it is a two-sided calculation.  Each tail gets half of alpha, so the confidence interval covers `confdence_level` and is centered on the mean.  See, e.g., step 2 with a known σ in [Confidence interval / Basic steps](https://en.wikipedia.org/wiki/Confidence_interval#Basic_steps).

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+
+def relative_risk(exposed_cases, exposed_total, control_cases, control_total):
+    """
+    Compute the relative risk (also known as the risk ratio).
```
The rendered documentation is very good (aside from `RelativeRiskResult` documentation not rendering)

mdhaber commented on scipy/stats/_relative_risk.py:
```diff
@@ -0,0 +1,266 @@
+    Find the confidence interval for the relative risk.
+
+    >>> result.confidence_interval(confidence_level=0.95)
+    ConfidenceInterval(low=1.5836990926700116, high=3.7886786315466354)
```
Checked. Example 3.2 reports the logs of these.

mdhaber commented on scipy/stats/contingency.py:
```diff
@@ -1,13 +1,29 @@
-"""Some functions for working with contingency tables (i.e. cross tabulations).
+"""
```
Looks like this was intended to be rendered, but I can't find it.

WarrenWeckesser replied:
I updated the module docstring in `contingency.py` to make Sphinx happy, but in fact we don't generate a separate page for the `contingency` submodule.  I don't think we need one, since everything is already covered in the main `stats` module docstring.

mdhaber commented on scipy/stats/contingency.py:
```diff
@@ -1,13 +1,29 @@
+   chi2_contingency
+   expected_freq
+   margins
+   relative_risk
```
Why no fisher_exact?

mdhaber replied:
Never mind, I see that `fisher_exact` is not actually in the namespace. Should it be?

mdhaber commented on scipy/stats/tests/test_relative_risk.py:
```diff
@@ -0,0 +1,96 @@
+    #
+    # The last line is the result that we want.
+    assert_allclose(rr, 4/3)
+    assert_allclose((ci.low, ci.high), (0.7347317, 2.419628), rtol=5e-7)
```
Confirmed against R.

mdhaber reviewed (approved):
Minor comments. Indicating approval now, but the codecov hints (e.g. line 104, 247, 249) are good, and the documentation of the `RelativeRiskResult` object should render before merging. I'd also appreciate responses to the other comments to confirm my understanding.

rlucas7 reviewed (commented):
I went through and marked my comments resolved. Also, I merged the `crosstab` pr that Warren had open so the module had 2 (1-line) conflicts. It looks like github allows us to resolve the conflicts so I gave that a go and pushed the changes here. From my end looks good but will wait for @mdhaber to have his comments addressed.

rlucas7 commented:
> I understand the motivation for making this a separate function from `fisher_exact`, taking the input arguments separately, and taking the totals rather than non-cases, but I still think it would be nice if `fisher_exact` could be augmented with relative risk, explaining the assumption about the order of the data in the docstring. But since this probably stems from our different perspectives on when to create a separate function (e.g. different `minimize` methods), which we've already discussed a few times, I won't say anything more here if @rlucas7 and @rgommers are OK with it.

I think of these as 2 different techniques and it seems weird to me to put them in the same function. They're different ways to getting to a test statistic or confidence interval, the first is a permutation test the second is a delta (approximation) method-you'd usually do one or the other but not both, at least not that I've seen anyway.

mdhaber commented:
Yeah I talked with @WarrenWeckesser and I retract that comment about adding to `fisher_exact`. I think I just want some consistent way of entering data for all these different contingency table tests, but adding this technique to `fisher_exact` is not the right way to do that.

rlucas7 commented:
I think that ship has sailed-maybe it could be left where it is and an alias inside contingency could be used? 

That might make sense.

Warren posted a link in one of this PRs but I don’t have it handy. I think it was PR 4x in this github repo, somewhere IN The first 100 IIRC, in case you want to read the discussion.

-Lucas

> On Dec 4, 2020, at 6:12 PM, Matt Haberland <notifications@github.com> wrote:
> 
> ﻿
> @mdhaber commented on this pull request.
> 
> In scipy/stats/contingency.py:
> 
> > -"""Some functions for working with contingency tables (i.e. cross tabulations).
> +"""
> +Contingency table functions (:mod:`scipy.stats.contingency`)
> +============================================================
> +
> +Functions for creating and analyzing contingency tables.
> +
> +.. currentmodule:: scipy.stats.contingency
> +
> +.. autosummary::
> +   :toctree: generated/
> +
> +   chi2_contingency
> +   expected_freq
> +   margins
> +   relative_risk
> Never mind, I see that fisher_exact is not actually in the namespace. Should it be?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

WarrenWeckesser commented:
I merged from master to resolve conflicts.  I think I have addressed all the requested changes.

WarrenWeckesser commented:
@rgommers and @rlucas7, I have updated the code and I think I have addressed your requests.  Let me know if there are still changes that you'd like to see.

mdhaber reviewed (approved):
Confirming my approval after changes. An email was sent to the mailing list "Adding the function `relative_risk` to SciPy" and it looks like there were no objections (no response at all, apparently). Two things for consideration, though:

- [ ] I still don't see mention of this being Method C of Katz' paper
- [x] This adds "Objects returned by functions in the contingency submodule" to the top level stats documentation. There has been discussion about this being undesirable. If there's another way of getting the `RelativeRiskResult` documentation to render (linked just from the `relative_risk` function), that may be preferable.

mdhaber commented:
@rgommers you gave thumbs up [above](https://github.com/scipy/scipy/pull/13048#issuecomment-768799596). Did that mean you approve? I ask because there has not been success implementing your [suggestion](https://github.com/scipy/scipy/issues/13118#issuecomment-770403118) from gh-13118 about rendering hypothesis test result class documentation:

> Even if we do it on a separate page, we can convince Sphinx to render that without putting it in the public API. 

How do we convince Sphinx to render it without putting it on the top-level stats documentation page?

rgommers commented:
> @rgommers you gave thumbs up [above](https://github.com/scipy/scipy/pull/13048#issuecomment-768799596). Did that mean you approve? I ask because there has not been success implementing your [suggestion](https://github.com/scipy/scipy/issues/13118#issuecomment-770403118) from [gh-13118](https://github.com/scipy/scipy/issues/13118) about rendering hypothesis test result class documentation:

That thumbs up was about adding `contingency.relative_risk` (mailing list discussion happened, and it's in the right subsubmodule). I didn't look at the rest of this PR in detail.

This indeed looks wrong:

<img width="422" alt="image" src="https://user-images.githubusercontent.com/98330/109388133-84846a80-7905-11eb-86db-7b6fc41ce484.png">

> How do we convince Sphinx to render it without putting it on the top-level stats documentation page?

Using the `:hidden:` or `:includehidden:` attributes to `:toctree:` should work I'd think: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html

rlucas7 reviewed (approved):
Ran tests locally in docker and tested out in repl: 
```
(scipydev) root@83b14aec8890:/home/scipy# python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy.stats
>>> import scipy.stats as ss
>>> import scipy.stats.contingency as ssc
>>> ssc.relative_risk(1,1,1,1)
RelativeRiskResult(relative_risk=1.0, exposed_cases=1, exposed_total=1, control_cases=1, control_total=1)
>>> a = ssc.relative_risk(1,1,1,1)
>>> a
RelativeRiskResult(relative_risk=1.0, exposed_cases=1, exposed_total=1, control_cases=1, control_total=1)
>>> type(a)
<class 'scipy.stats._relative_risk.RelativeRiskResult'>
>>> a.control_total
1
>>> a.
a.confidence_interval(  a.control_cases         a.control_total         a.exposed_cases         a.exposed_total         a.relative_risk         
>>> help(a.confidence_interval)

>>> a = ssc.relative_risk(10,75,12,225)
>>> a.confidence_interval(0.95)
ConfidenceInterval(low=1.1261564003469628, high=5.549850800541033)
>>> help(a.confidence_interval)

>>> a.relative_risk
2.5

>>> a = ssc.relative_risk(10,750,0,225)
>>> a.relative_risk
inf
>>> a = ssc.relative_risk(0,750,5,225)
>>> a.relative_risk
0.0
>>> a = ssc.relative_risk(0,0,5,225)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/scipy/scipy/stats/_relative_risk.py", line 233, in relative_risk
    exposed_total = _validate_int(exposed_total, 1, "exposed_total")
  File "/home/scipy/scipy/stats/_relative_risk.py", line 16, in _validate_int
    raise ValueError(msg)
ValueError: exposed_total must be an integer not less than 1, but got 0
>>> a = ssc.relative_risk(0,1,0,225)
>>> a.relative_risk
nan
>>> 
```
things seem to behave as expected. Also, I built and checked the docs:

```
120 <tr class="row-odd"><td><p><a class="reference internal" href="generated/scipy.stats.contingency.relative_risk.html#scipy.stats.contingency.relative_risk" title="scipy.stats.co     ntingency.relative_risk"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contingency.relative_risk</span></code></a>(exposed_cases,?| ?~@?)</p></td>
1121 <td><p>Compute the relative risk (also known as the risk ratio).</p></td>       
1122 </tr>           
```
it's difficult for me to check the built docs inside the docker container but this seeems to be what we want.

I'm marking approved and if no further comments by March 7th, I'll merge.

mdhaber commented:
> I see the note in the docstring, it's line 61 of this file

Yes it seems to have been added after my note.

One thing to do before merging is fix the docs as Ralf suggested.

rlucas7 commented on scipy/stats/__init__.py:
```diff
@@ -432,7 +433,17 @@
+==================================
+
+.. autosummary::
+   :toctree: generated/
```
```suggestion
   :toctree: generated/
   :hidden:
```

rlucas7 replied:
>One thing to do before merging is fix the docs as Ralf suggested.

Not sure I follow what we want to achieve w.r.t. the docs @mdhaber does this match? 

IIUC, that will ***hide*** the `scipy.stats._result_classes.RelativeRiskResult` entry from the TOC in sphinx.

mdhaber replied:
Yes, I think so. We would also want to remove the heading of this new section, which would no longer exist.

rlucas7 reviewed (commented):
@mdhaber I left a suggestion, can you check if that matches what we're trying to achieve? 
I'm not sure I follow what we want w.r.t. docs.

rlucas7 commented:
> On Mar 1, 2021, at 1:18 AM, Matt Haberland <notifications@github.com> wrote:
> 
> ﻿
> @mdhaber commented on this pull request.
> 
> In scipy/stats/__init__.py:
> 
> > @@ -432,7 +433,17 @@
>     PearsonRNearConstantInputWarning
>     SpearmanRConstantInputWarning
>  
> +
> +Classes returned by some functions
> +==================================
> +
> +.. autosummary::
> +   :toctree: generated/
> Yes, I think so. We would also want to remove the heading of this new section, which would no longer exist.
> 

I might’ve missed another PR. 

The BinomTestResult is already in that section on master? 

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mdhaber commented:
> The BinomTestResult is already in that section on master?

Oh, I guess so. But we want the same thing for that - we don't want that section of the top level documentation to exist. We want the documentation of those statistical test result classes to be linked only from the documentation of the statistical test functions.

WarrenWeckesser commented:
Is there a way to view the Circle CI error details without creating a Circle CI account?  When I click on the "Details" link for `ci/circleci: build_docs`, Circle CI asks me to log in.

rlucas7 commented:
> Is there a way to view the Circle CI error details without creating a Circle CI account? When I click on the "Details" link for `ci/circleci: build_docs`, Circle CI asks me to log in.

I ran into the same thing awhile back, the only way i was able to work around it was to setup an account.

rlucas7 commented:
> > Is there a way to view the Circle CI error details without creating a Circle CI account? When I click on the "Details" link for `ci/circleci: build_docs`, Circle CI asks me to log in.
> 
> I ran into the same thing awhile back, the only way i was able to work around it was to setup an account.

@WarrenWeckesser 
Here's what I see in circle ci:

```
#!/bin/bash -eo pipefail
. venv/bin/activate
export SHELL=$(which bash)
python -u runtests.py -g -j2 --doc html-scipyorg --doc latex
make -C doc/build/latex all-pdf LATEXOPTS="-file-line-error -halt-on-error"
cp -f doc/build/latex/scipy-ref.pdf doc/build/html-scipyorg/

Building, see build.log...
Build OK (0:00:31.601289 elapsed)
make: Entering directory '/home/circleci/repo/doc'
mkdir -p build/html build/doctrees
LANG=C "/home/circleci/repo/venv/bin/python" -msphinx -WT --keep-going -t scipyorg  -b html -d build/doctrees  "-j2" source build/html-scipyorg 
Running Sphinx v3.5.1
SciPy (VERSION 1.7.0.dev0+eedbab1)
making output directory... done
[autosummary] generating autosummary for: api.rst, ccallback.rst, cluster.hierarchy.rst, cluster.rst, cluster.vq.rst, constants.rst, fft.rst, fftpack.rst, hacking.rst, index.rst, ..., sparse.rst, spatial.distance.rst, spatial.rst, spatial.transform.rst, special.cython_special.rst, special.rst, stats.mstats.rst, stats.qmc.rst, stats.rst, toolchain.rst
[autosummary] generating autosummary for: /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.ClusterNode.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.DisjointSet.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.average.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.centroid.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.complete.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.cophenet.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.correspond.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.cut_tree.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.dendrogram.rst, ..., /home/circleci/repo/doc/source/generated/scipy.stats.wrapcauchy.rst, /home/circleci/repo/doc/source/generated/scipy.stats.yeojohnson.rst, /home/circleci/repo/doc/source/generated/scipy.stats.yeojohnson_llf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.yeojohnson_normmax.rst, /home/circleci/repo/doc/source/generated/scipy.stats.yeojohnson_normplot.rst, /home/circleci/repo/doc/source/generated/scipy.stats.yulesimon.rst, /home/circleci/repo/doc/source/generated/scipy.stats.zipf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.zipfian.rst, /home/circleci/repo/doc/source/generated/scipy.stats.zmap.rst, /home/circleci/repo/doc/source/generated/scipy.stats.zscore.rst
[autosummary] generating autosummary for: /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.__getitem__.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.__len__.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.__mul__.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.count.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.from_cython.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.function.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.index.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.signature.rst, /home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.user_data.rst, /home/circleci/repo/doc/source/generated/scipy.cluster.hierarchy.ClusterNode.get_count.rst, ..., /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.nnlf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.pdf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.ppf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.random_state.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.rvs.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.sf.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.stats.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.std.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.support.rst, /home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.var.rst
loading intersphinx inventory from https://docs.python.org/dev/objects.inv...
loading intersphinx inventory from https://numpy.org/devdocs/objects.inv...
loading intersphinx inventory from https://numpy.org/neps/objects.inv...
loading intersphinx inventory from https://matplotlib.org/objects.inv...
loading intersphinx inventory from https://asv.readthedocs.io/en/stable/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 294 source files that are out of date
updating environment: [new config] 3603 added, 0 changed, 0 removed
reading sources... [  3%] api .. generated/scipy.fft.ifftshift
reading sources... [  7%] generated/scipy.fft.ihfft .. generated/scipy.integrate.fixed_quad
reading sources... [ 11%] generated/scipy.integrate.newton_cotes .. generated/scipy.interpolate.LSQSphereBivariateSpline.get_residual
reading sources... [ 14%] generated/scipy.interpolate.LSQUnivariateSpline .. generated/scipy.io.arff.MetaData.__getitem__
... many more of these ...
reading sources... [ 81%] generated/scipy.special.obl_rad1_cv .. generated/scipy.stats.chisquare
reading sources... [ 85%] generated/scipy.stats.circmean .. generated/scipy.stats.mstats.gmean
reading sources... [ 88%] generated/scipy.stats.mstats.hdmedian .. generated/scipy.stats.rankdata
reading sources... [ 92%] generated/scipy.stats.ranksums .. generated/scipy.stats.yeojohnson_normplot
reading sources... [ 96%] generated/scipy.stats.yulesimon .. tutorial/index
reading sources... [100%] tutorial/integrate .. tutorial/stats/discrete_zipfian

waiting for workers...
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/__init__.py:docstring of scipy.stats:431: WARNING: Error in "autosummary" directive:
unknown option: "hidden".

.. autosummary::
   :toctree: generated/
   :hidden:

   scipy.stats._result_classes.BinomTestResult
   scipy.stats._result_classes.RelativeRiskResult
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/circleci/repo/doc/source/generated/scipy.stats._result_classes.BinomTestResult.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats._result_classes.RelativeRiskResult.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [  5%] building/faq .. generated/scipy.integrate.BDF.step
writing output... [ 10%] generated/scipy.integrate.BDF.step_size .. generated/scipy.interpolate.InterpolatedUnivariateSpline.get_knots
writing output... [ 15%] generated/scipy.interpolate.InterpolatedUnivariateSpline.get_residual .. generated/scipy.linalg.LinAlgWarning
writing output... [ 20%] generated/scipy.linalg.blas.caxpy .. generated/scipy.linalg.interpolative.reconstruct_matrix_from_id
writing output... [ 25%] generated/scipy.linalg.interpolative.reconstruct_skel_matrix .. generated/scipy.linalg.lapack.dgesvx
writing output... [ 30%] generated/scipy.linalg.lapack.dgetc2 .. generated/scipy.linalg.lapack.spotrs
writing output... [ 35%] generated/scipy.linalg.lapack.sppcon .. generated/scipy.linalg.lapack.ztgsen
writing output... [ 40%] generated/scipy.linalg.lapack.ztgsen_lwork .. generated/scipy.optimize.LbfgsInvHessProduct.adjoint
writing output... [ 45%] generated/scipy.optimize.LbfgsInvHessProduct.dot .. generated/scipy.signal.gammatone
writing output... [ 50%] generated/scipy.signal.gauss_spline .. generated/scipy.sparse.bsr_matrix.sin
writing output... [ 55%] generated/scipy.sparse.bsr_matrix.sinh .. generated/scipy.sparse.csgraph.structural_rank
writing output... [ 60%] generated/scipy.sparse.csr_matrix .. generated/scipy.sparse.dok_matrix.tocsr
writing output... [ 65%] generated/scipy.sparse.dok_matrix.todense .. generated/scipy.spatial.HalfspaceIntersection.close
writing output... [ 70%] generated/scipy.spatial.HalfspaceIntersection.dual_vertices .. generated/scipy.special.erfcinv
writing output... [ 75%] generated/scipy.special.erfcx .. generated/scipy.special.pro_rad1
writing output... [ 80%] generated/scipy.special.pro_rad1_cv .. generated/scipy.stats.invgamma
writing output... [ 85%] generated/scipy.stats.invgauss .. generated/scipy.stats.randint
writing output... [ 90%] generated/scipy.stats.random_correlation .. optimize.root-diagbroyden
writing output... [ 95%] optimize.root-excitingmixing .. tutorial/stats/discrete_hypergeom
writing output... [100%] tutorial/stats/discrete_logser .. tutorial/stats/discrete_zipfian

waiting for workers...
generating indices... genindex py-modindex done
writing additional pages... search done
copying images... [  0%] dev/gitwash/forking_button.png
copying images... [  0%] dev/gitwash/pull_button.png
copying images... [  0%] ../build/plot_directive/generated/scipy-cluster-hierarchy-dendrogram-1_00.png
copying images... [  0%] ../build/plot_directive/generated/scipy-cluster-hierarchy-dendrogram-1_01.png
copying images... [  0%] ../build/plot_directive/generated/scipy-cluster-hierarchy-leaves_list-1.png
copying images... [  1%] ../build/plot_directive/generated/scipy-cluster-hierarchy-linkage-1_00.png
... many more of these ...
copying images... [ 99%] ../build/plot_directive/tutorial/stats/plots/kde_plot5.png
copying images... [ 99%] ../build/plot_directive/tutorial/stats/plots/mgc_plot1_01_00.png
copying images... [ 99%] ../build/plot_directive/tutorial/stats/plots/mgc_plot2.png
copying images... [ 99%] ../build/plot_directive/tutorial/stats/plots/mgc_plot3_01_00.png
copying images... [100%] ../build/plot_directive/tutorial/stats/plots/mgc_plot4.png

copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 3 warnings.
Makefile:127: recipe for target 'html-scipyorg' failed
make: *** [html-scipyorg] Error 1
make: Leaving directory '/home/circleci/repo/doc'
Traceback (most recent call last):
  File "runtests.py", line 565, in <module>
    main(argv=sys.argv[1:])
  File "runtests.py", line 214, in main
    subprocess.run(cmd, check=True)
  File "/usr/local/lib/python3.7/subprocess.py", line 468, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['make', '-Cdoc', 'PYTHON="/home/circleci/repo/venv/bin/python"', 'html-scipyorg', 'latex', 'SPHINXOPTS="-j2"']' returned non-zero exit status 2.

Exited with code exit status 1

CircleCI received exit code 1
```

Lookes like the `Makefile` LIC 127 is the error source.

WarrenWeckesser commented:
> I'm wondering if one of the flags on line 128 makes a warning error out?

You're right, that's what is happening.  The flag `-W` turns warnings into errors.  The new warning is about `autosummary` not having the `:hidden:` option.  Running `make html-scipyorg` returns an error (even though the docs appear to be built successfully).

I'm about to push an update that uses a different approach to getting the test result classes into the HTML docs, in which this `autosummary` directive is removed completely.

rlucas7 commented:
> > I'm wondering if one of the flags on line 128 makes a warning error out?
> 
> You're right, that's what is happening. The flag `-W` turns warnings into errors. The new warning is about `autosummary` not having the `:hidden:` option. Running `make html-scipyorg` returns an error (even though the docs appear to be built successfully).

Yeah that's what I was thinking but wasn't sure about the specifics of the `T` next to the `-W`-thanks for clarifying. 
 
> I'm about to push an update that uses a different approach to getting the test result classes into the HTML docs, in which this `autosummary` directive is removed completely.

Not sure if you saw this already but it looks like the `:autosummary:` approach results in warnings too:

https://app.circleci.com/pipelines/github/scipy/scipy/5714/workflows/fa0ce4e6-ab2c-4b19-bdce-371162b623f4/jobs/27222/parallel-runs/0/steps/0-103?invite=true

in particular these lines: 

```
waiting for workers...
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/_binomtest.py:docstring of scipy.stats._binomtest.BinomTestResult.rst:38: WARNING: autosummary: stub file not found 'scipy.stats._result_classes.BinomTestResult.proportion_ci'. Check your autosummary_generate setting.
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/_relative_risk.py:docstring of scipy.stats._relative_risk.RelativeRiskResult.rst:40: WARNING: autosummary: stub file not found 'scipy.stats._result_classes.RelativeRiskResult.confidence_interval'. Check your autosummary_generate setting.
looking for now-outdated files... none found
```

I'm not sure if this will work (I found it but didn't try it yet): 
https://stackoverflow.com/a/65203035/3164100
seems like similar/same warning here in circleci, or it could be a config for sphinx but I'm not sure if/where we keep those in the repo. 

It looks like the sphinx configs are empty by default:
https://github.com/scipy/scipy/blob/master/doc/Makefile#L8
is that maybe what needs to be changed to not raise a warning? 

O/w I'm out of ideas.

mdhaber commented:
I'm not sure if it helps, but I used `:hidden:` before in [`index.rst`](https://github.com/scipy/scipy/blob/master/doc/source/index.rst). I also had to mark the things it refers to as `:orphan:`, e.g. [here](https://github.com/scipy/scipy/blob/master/doc/source/dev/contributor/quickstart_mac.rst). Is it worth trying:
```
.. toctree::
   :hidden:
```
instead of
```
.. autosummary::
   :toctree: generated/
   :hidden:
```?

I haven't read the conversation above very carefully, and I really don't know how Sphinx is supposed to work, so never mind if it's not so simple.

I would try myself, but I have never been able to build the documentation on Windows.

WarrenWeckesser commented:
@mdhaber, we can't replace `autosummary` with `toctree` because (at least as far as I understand it) the items listed under the  `toctree` directive are .rst files, while the items under the `autosummary` directive are Python names to be processed by the autodoc extension.

WarrenWeckesser commented:
@rlucas7, thanks, the links provided a suggestion that fixed a problem.  In the result class docstrings, I had taken out the short blurbs for methods listed in the "Methods" section, and in the rendered HTML, Sphinx would automatically add the short blurb and would make the method name a link to the rendered docs for the method.  If the blurb is given explicitly (as it is now), Sphinx doesn't convert the method name into a link.  This might be acceptable for now, since, at least for the two classes in this PR, there is just one method, and it is rendered immediately below the class docs.

Anyway, the warnings are fixed.  With the latest version of this PR, all the result classes are rendered on the same [HTML page](https://27252-1460385-gh.circle-artifacts.com/0/html-scipyorg/stats._result_classes.html).  The links to the result classes in the [`binomtest`](https://27252-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.stats.binomtest.html) and [`relative_risk`](https://27252-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.stats.contingency.relative_risk.html) docstrings go to the class docs within that page.

mdhaber reviewed (approved):
Looks fine for now. I think that it is confusing to have all of that on the same page but that can be a separate discussion.

rlucas7 commented:
> Looks fine for now. I think that it would be better to split the documentation for those classes into separate pages, but that can be a separate discussion.

I agree. 
 
> ... so I'll merge Friday afternoon unless there is new feedback to address.

@mdhaber Sounds good I'll leave it to you to merge then

mdhaber commented:
Thanks @WarrenWeckesser @rlucas7 @rgommers!
